massa-execution/src/vm.rs

```diff
@@ -174,6 +177,63 @@ impl VM {
         context.ledger_step.caused_changes.clone()
     }
 
+    /// Run code in read-only mode
+    pub(crate) fn run_read_only(
+        &self,
+        slot: Slot,
+        max_gas: u64,
+        simulated_gas_price: Amount,
+        bytecode: Vec<u8>,
+        address: Option<Address>,
+        result_sender: oneshot::Sender<ExecuteReadOnlyResponse>,
+    ) {
+        let saved_changes = {
+            let mut context = self.execution_context.lock().unwrap();
+
+            // Set the call stack, using the provided address, or a random one.
+            let address = address.unwrap_or_else(|| {
+                let private_key = generate_random_private_key();
+                let public_key = derive_public_key(&private_key);
+                Address::from_public_key(&public_key)
+            });
+            context.call_stack = vec![address].into();
+
+            // Set the max gas.
+            context.max_gas = max_gas;
+
+            // Set the simulated gas price.
+            context.gas_price = simulated_gas_price;
+
+            // Return a clone of the current state, to revert to it after execution.
+            context.ledger_step.caused_changes.clone()
+        };
+
+        // run in the intepreter
+        let run_result = assembly_simulator::run(&bytecode, max_gas, &*self.execution_interface);
+
+        // cancel the effects of execution only, pop back init_changes
+        self.execution_context
+            .lock()
+            .unwrap()
+            .ledger_step
+            .caused_changes = saved_changes;
+
+        // Send result back.
+        let execution_response = ExecuteReadOnlyResponse {
+            executed_at: slot,
+            // TODO: specify result.
+            result: run_result.map_or_else(
+                |_| ReadOnlyResult::Error("Failed to run in read-only mode".to_string()),
+                |_| ReadOnlyResult::Ok,
+            ),
```

Comment:
We can return all the response or at least the `run_result.res`, what do you think?

Reply:
I'm not sure, do we actually want to return anything other than the output events? @damip Looking at the latest version at https://github.com/massalabs/massa/discussions/1970#discussioncomment-1879779 it seems we don't.